python/taichi/lang/util.py

```diff
@@ -170,10 +170,35 @@ def decorator(foo):
         def wrapped(*args, **kwargs):
             import warnings
             msg = f'{old} is deprecated, please use {new} instead'
-            #warnings.warn(msg, DeprecationWarning)
-            raise DeprecationWarning(msg)
+            warnings.warn(msg, DeprecationWarning)
             return foo(*args, **kwargs)
 
         return wrapped
 
     return decorator
+
+
+def taichi_scope(func):
+    import functools
+    from . import impl
+
+    @functools.wraps(func)
+    def wrapped(*args, **kwargs):
+        assert impl.inside_kernel(), \
```

Comment:
FYI, according to #1066, I believe it's better to throw an exception here, since we cannot control how user would use these operators.

Reply:
I will stay as-is, since currently we already have many `assert impl.inside_kernel()`, and it's wise to fix all these issues in another single PR. Also note that removing these assertions when `python -O` is imo OK for performance.